### PR TITLE
fix relative motion support

### PR DESCRIPTION
**Pull request by jafu888**
Still need cleanup and  
image
more testing

### CODE REVIEW

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/motion/Motion.java [hunk 1]
```diff
@@ -237,16 +239,23 @@ public float getFinalHeight() {
...
     /**
-     * @TODO: add description
+     * set up the motion to be relative to this other motionController
      */
     public void setupRelative(Motion motionController) {
-        mStartMotionPath.setupRelative(motionController, motionController.mStartMotionPath);
-        mEndMotionPath.setupRelative(motionController, motionController.mEndMotionPath);
+        mRelativeMotion = motionController;
+    }
+    private void setupRelative() {
```
Comment (oscar-ad) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/motion/Motion.java:253]: Needs a line in between

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/motion/Motion.java [hunk 2]
```diff
@@ -1722,7 +1734,10 @@ public boolean setValue(int id, String value) {
             mQuantizeMotionInterpolator = getInterpolator(SPLINE_STRING, value, 0);
             return true;
         }
-
+        if ( MotionType.TYPE_ANIMATE_RELATIVE_TO == id) {
```
Comment (oscar-ad) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/motion/Motion.java:1737]: Fix spacing:
`if (MotionType...`

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/motion/Motion.java [hunk 3]
```diff
@@ -1763,4 +1778,9 @@ public void setStaggerOffset(float staggerOffset) {
     public float getMotionStagger() {
         return mMotionStagger;
     }
+
+    public void setIdString(String stringId) {
+        mId = stringId;
+        mStartMotionPath.mId = mId;
```
Comment (oscar-ad) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/motion/Motion.java:1784]: What about `mEndMotionPath.mId = mId`?
Is it not relevant there?

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/motion/MotionPaths.java [hunk 4]
```diff
@@ -157,6 +158,9 @@ void initPolar(int parentWidth,
         mHeight = (int) (s.mHeight + scaleY * scaleHeight);
         float startfactor = 1 - position;
         float endfactor = position;
+        Utils.log(mId+": start "+s.mX+","+s.mY);
+        Utils.log(mId+": end "+e.mX+","+e.mY);
+        Utils.log(mId+": mAnimateRelativeTo "+e.mAnimateRelativeTo);
```
Comment (oscar-ad) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/motion/MotionPaths.java:163]: Fix spacing between operators.

File: constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/Transition.java [hunk 5]
```diff
@@ -710,11 +710,31 @@ public void addCustomColor(int state, String widgetId, String property, int colo
...
+        WidgetState [] states = new WidgetState[count];
+        boolean relative = false;
         for (int i = 0; i < count; i++) {
             ConstraintWidget child = children.get(i);
             WidgetState widgetState = getWidgetState(child.stringId, null, state);
+            states[i] = widgetState;
             widgetState.update(child, state);
+           String id =  widgetState.getPathRelativeId();
+            if (id != null) {
+                relative = true;
```
Comment (oscar-ad) [inline constraintlayout/core/src/main/java/androidx/constraintlayout/core/state/Transition.java:722]: Seems like we could just have `widgetState.setPathRelative(getWidgetState(id, null, state));` here instead.
No need to loop twice.

**Review by oscar-ad — CHANGES_REQUESTED** [review]

**Comment by jswong65** [comment]
Core / Assemble checks failed. Could you please fix it
```
/home/runner/work/constraintlayout/constraintlayout/constraintlayout/core/src/main/java/androidx/constraintlayout/core/motion/Motion.java:34: error: cannot find symbol
import androidx.constraintlayout.core.motion.utils.Nullable;
> Task :core:compileJava
                                                  ^
  symbol:   class Nullable
  location: package androidx.constraintlayout.core.motion.utils
/home/runner/work/constraintlayout/constraintlayout/constraintlayout/core/src/main/java/androidx/constraintlayout/core/motion/Motion.java:241: error: cannot find symbol
    @Nullable
     ^
  symbol:   class Nullable
  location: class Motion
```

https://github.com/androidx/constraintlayout/runs/6941085866?check_suite_focus=true

**Review by jswong65 — APPROVED** [review]

**Pull request merged**